### ISSUE TITLE
Bug: Gluetun keeps saying "credentials might be wrong" / Privado

### ISSUE THREAD

**Issue by MichaelMichaelMichaelMichaelMichael** [qdm12/gluetun#1178]
### Is this urgent?

No

### Host OS

Ubuntu 22.04.1 LTS

### CPU arch

x86_64

### VPN service provider

Privado

### What are you using to run the container

Portainer

### What is the version of Gluetun

v3.31.1 (latest)

### What's the problem 🤔

I cannot login into the vpn server, gluetun says: "Your credentials might be wrong 🤨" (see logs section)

What works, as a control:
commandline openvpn with the provided files by Privado, I can connect after slightly modifying them (the stock files don't work for me), e.g. 
**gru-003.ovpn:**
```
client
dev tun
proto udp
remote gru-003.vpn.privado.io 1194
resolv-retry infinite
nobind
persist-key
persist-tun
persist-remote-ip

verify-x509-name gru-003.vpn.privado.io name

tls-client

route-ipv6 ::/0
route 0.0.0.0 0.0.0.0 vpn_gateway
tls-cipher TLS-DHE-RSA-WITH-AES-256-CBC-SHA:TLS-DHE-DSS-WITH-AES-256-CBC-SHA:TLS-RSA-WITH-AES-256-CBC-SHA
remote-cert-tls server
auth-user-pass login.conf
verb 3
auth SHA256
route-delay 3

<ca>
-----BEGIN CERTIFICATE-----
...
-----END CERTIFICATE-----
</ca>

```

**login.conf:**
```
vpnusername
vpnpassword
```

command to connect manually via cli:

```
$ sudo openvpn --config gru-003.ovpn
...
Sat Oct  1 17:13:12 2022 OpenVPN 2.4.7 x86_64-pc-linux-gnu [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [PKCS11] [MH/PKTINFO] [AEAD] built on Mar 22 2022
Sat Oct  1 17:13:12 2022 library versions: OpenSSL 1.1.1f  31 Mar 2020, LZO 2.10
...
Sat Oct  1 17:13:12 2022 VERIFY OK: depth=1, CN=Privado
Sat Oct  1 17:13:12 2022 VERIFY KU OK
Sat Oct  1 17:13:12 2022 Validating certificate extended key usage
Sat Oct  1 17:13:12 2022 ++ Certificate has EKU (str) TLS Web Server Authentication, expects TLS Web Server Authentication
Sat Oct  1 17:13:12 2022 VERIFY EKU OK
Sat Oct  1 17:13:12 2022 VERIFY X509NAME OK: CN=gru-003.vpn.privado.io
Sat Oct  1 17:13:12 2022 VERIFY OK: depth=0, CN=gru-003.vpn.privado.io
...
Sat Oct  1 17:13:13 2022 Control Channel: TLSv1.2, cipher SSLv3 DHE-RSA-AES256-SHA, 4096 bit RSA
Sat Oct  1 17:13:13 2022 [gru-003.vpn.privado.io] Peer Connection Initiated with [AF_INET]129.227.121.233:1194
Sat Oct  1 17:13:14 2022 SENT CONTROL [gru-003.vpn.privado.io]: 'PUSH_REQUEST' (status=1)
...
Sat Oct  1 17:13:18 2022 Initialization Sequence Completed

```

### Share your logs

```plain text
|   |           └── Protocol: UDP
|   └── OpenVPN settings:
|       ├── OpenVPN version: 2.5
|       ├── User: [set]
|       ├── Password: [set]
|       ├── Tunnel IPv6: no
|       ├── Network interface: tun0
|       ├── Run OpenVPN as: root
|       └── Verbosity level: 1
├── DNS settings:
|   ├── DNS server address to use: 127.0.0.1
|   ├── Keep existing nameserver(s): no
|   └── DNS over TLS settings:
|       ├── Enabled: yes
|       ├── Update period: every 24h0m0s
|       ├── Unbound settings:
|       |   ├── Authoritative servers:
|       |   |   └── cloudflare
|       |   ├── Caching: yes
|       |   ├── IPv6: no
|       |   ├── Verbosity level: 1
|       |   ├── Verbosity details level: 0
|       |   ├── Validation log level: 0
|       |   ├── System user: root
|       |   └── Allowed networks:
|       |       ├── 0.0.0.0/0
|       |       └── ::/0
|       └── DNS filtering settings:
|           ├── Block malicious: yes
|           ├── Block ads: no
|           ├── Block surveillance: no
|           └── Blocked IP networks:
|               ├── 127.0.0.1/8
|               ├── 10.0.0.0/8
|               ├── 172.16.0.0/12
|               ├── 192.168.0.0/16
|               ├── 169.254.0.0/16
|               ├── ::1/128
|               ├── fc00::/7
|               ├── fe80::/10
|               ├── ::ffff:7f00:1/104
|               ├── ::ffff:a00:0/104
|               ├── ::ffff:a9fe:0/112
|               ├── ::ffff:ac10:0/108
|               └── ::ffff:c0a8:0/112
├── Firewall settings:
|   └── Enabled: yes
├── Log settings:
|   └── Log level: DEBUG
├── Health settings:
|   ├── Server listening address: 127.0.0.1:9999
|   ├── Target address: cloudflare.com:443
|   ├── Read header timeout: 100ms
|   ├── Read timeout: 500ms
|   └── VPN wait durations:
|       ├── Initial duration: 6s
|       └── Additional duration: 5s
├── Shadowsocks server settings:
|   └── Enabled: no
├── HTTP proxy settings:
|   └── Enabled: no
├── Control server settings:
|   ├── Listening address: :8000
|   └── Logging: yes
├── OS Alpine settings:
|   ├── Process UID: 1000
|   ├── Process GID: 1000
|   └── Timezone: Europe/Berlin
├── Public IP settings:
|   ├── Fetching: every 12h0m0s
|   └── IP file path: /tmp/gluetun/ip
└── Version settings:
    └── Enabled: yes
2022-10-01T17:20:36+02:00 INFO [routing] default route found: interface eth0, gateway x.x.x.1 and assigned IP x.x.x.2
2022-10-01T17:20:36+02:00 DEBUG [routing] ip rule add from x.x.x.2/32 lookup 200 pref 100
2022-10-01T17:20:36+02:00 INFO [routing] adding route for 0.0.0.0/0
2022-10-01T17:20:36+02:00 DEBUG [routing] ip route replace 0.0.0.0/0 via x.x.x.1 dev eth0 table 200
2022-10-01T17:20:36+02:00 INFO [firewall] setting allowed subnets...
2022-10-01T17:20:36+02:00 INFO [routing] default route found: interface eth0, gateway x.x.x.1 and assigned IP x.x.x.2
2022-10-01T17:20:36+02:00 INFO TUN device is not available: open /dev/net/tun: no such file or directory; creating it...
2022-10-01T17:20:36+02:00 INFO [pprof] http server listening on [::]:6060
2022-10-01T17:20:36+02:00 INFO [http server] http server listening on [::]:8000
2022-10-01T17:20:36+02:00 INFO [healthcheck] listening on 127.0.0.1:9999
2022-10-01T17:20:36+02:00 INFO [dns over tls] using plaintext DNS at address 1.1.1.1
2022-10-01T17:20:36+02:00 INFO [firewall] allowing VPN connection...
2022-10-01T17:20:36+02:00 DEBUG [firewall] iptables --append OUTPUT -d 146.59.31.2 -o eth0 -p udp -m udp --dport 1194 -j ACCEPT
2022-10-01T17:20:36+02:00 DEBUG [firewall] iptables --append OUTPUT -o tun0 -j ACCEPT
2022-10-01T17:20:36+02:00 DEBUG [firewall] ip6tables --append OUTPUT -o tun0 -j ACCEPT
2022-10-01T17:20:36+02:00 INFO [openvpn] OpenVPN 2.5.6 x86_64-alpine-linux-musl [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [MH/PKTINFO] [AEAD] built on Apr 17 2022
2022-10-01T17:20:36+02:00 INFO [openvpn] library versions: OpenSSL 1.1.1q  5 Jul 2022, LZO 2.10
2022-10-01T17:20:36+02:00 INFO [openvpn] TCP/UDP: Preserving recently used remote address: [AF_INET]146.59.31.2:1194
2022-10-01T17:20:36+02:00 INFO [openvpn] UDP link local: (not bound)
2022-10-01T17:20:36+02:00 INFO [openvpn] UDP link remote: [AF_INET]146.59.31.2:1194
2022-10-01T17:20:37+02:00 INFO [openvpn] [waw-005.vpn.privado.io] Peer Connection Initiated with [AF_INET]146.59.31.2:1194
2022-10-01T17:20:38+02:00 ERROR [openvpn] AUTH: Received control message: AUTH_FAILED

Your credentials might be wrong 🤨


2022-10-01T17:20:38+02:00 INFO [openvpn] SIGUSR1[soft,auth-failure] received, process restarting
```

### Share your configuration

```yml
version: "3"
services:
  gluetun:
    container_name: gluetun
    image: qmcgaw/gluetun
    cap_add:
      - NET_ADMIN
    environment:
      - VPN_SERVICE_PROVIDER=privado
      - OPENVPN_USER='vpnusername'
      - OPENVPN_PASSWORD='vpnpassword'
      - SERVER_COUNTRIES=Brazil
      - SERVER_HOSTNAMES=gru-003.vpn.privado.io,gru-004.vpn.privado.io
      - TZ=Europe/Berlin
      - LOG_LEVEL=debug
```

**Comment by MichaelMichaelMichaelMichaelMichael** [comment]
I have succeeded in using a custom openvpn configuration in the above mentioned format, as documented [here](https://github.com/qdm12/gluetun/wiki/OpenVPN-configuration-file):

`docker run -it --rm --cap-add=NET_ADMIN -e VPN_SERVICE_PROVIDER=custom -e OPENVPN_CUSTOM_CONFIG=/gluetun/custom.conf -v /root/gluetun/custom.conf:/gluetun/custom.conf:ro -e OPENVPN_USER='vpnusername' -e OPENVPN_PASSWORD='vpnpassword' qmcgaw/gluetun
`

This works flawlessly as a workaround. An easier method would of course be much appreciated.

**Comment by MichaelMichaelMichaelMichaelMichael** [comment]
Solved it in the end - I had a character in the OPENVPN_PASSWORD which was interpreted by the character substituion of docker-compose. I've escaped that and thus solved the problem. Closing.

**Comment by qdm12** [comment]
Awesome thanks for letting me know the resolution.
You can always bind mount passwords as files/secret files as an alternative as well, in case env variables are troublesome.

**Comment by KernelDeimos** [comment]
For the next person to get here from an online search (if that's still a thing by then... maybe you're getting this from an LLMs training data), I found the information here to be incomplete. I also needed to know the following:
- The character that needs to be escaped is `$` because of variable substitution (referred to as "character substitution" above).
- The syntax for escaping is **not** conventional. Instead of `\$` it's `$$`.

If your password is `aaa$aaa`, write it as `aaa$$aaa`.
